File: leetcode/easy/104-maxDepth.js
// Given the root of a binary tree, return its maximum depth.

// A binary tree's maximum depth is the number of nodes along the longest path from the root node down to the farthest leaf node.

// Input: root = [3,9,20,null,null,15,7]
// Output: 3

// Time: O(n)
// Space: O(n)

const maxDepth = (root) => {
  // empty tree has a height of 0
  if (root === null) return 0
  const leftDepth = maxDepth(root.left)
  const rightDepth = maxDepth(root.right)
  return 1 + Math.max(leftDepth, rightDepth)
}

/*
- Use recursion so the answer bubbles up from the bottom
- Need to add 1 after each call so that the comparison moves up (or it'll just stay the same) 
*/

// maxDepth([a,b,c,null,null,d,e])

//     a
//   /   \
//  b     c
//       /  \
//      d     e


// b -> leftDepth = 0, rightDepth = 0 + 1 = 1
//     a
//   /   \
//  1     c
//       /  \
//      d     e

// d -> leftDepth = 0, rightDepth = 0 + 1 = 1
//     a
//   /   \
//  1     c
//       /  \
//     1     e

// e -> leftDepth = 0, rightDepth = 0 + 1 = 1
//     a
//   /   \
//  1     c
//       /  \
//     1      1

// c -> leftDepth = 1, rightDepth = 1 + 1 = 2
//     a
//   /   \
//  1     2
//       /  \
//     1      1


// a -> leftDepth = 1, rightDepth = 2 + 1 = 3
//     3
//   /   \
//  1     2
//       /  \
//     1      1

File: leetcode/easy/104-maxDepth.py
# DFS Recursive
def maxDepth(root):
  if not root:
    return 0

  return 1 + max(maxDepth(root.left), maxDepth(root.right))

# BFS Iterative
def maxDepth(root):
  if not root:
    return 0
  
  level = 0
  queue = deque([ root ])
  while queue:
    for i in range(len(queue)):
      node = queue.popleft()
      if node.left:
        queue.append(node.left)
      if node.right:
        queue.append(node.right)
    
    level += 1
  return level

# DFS Iterative - Pre-order
def maxDepth(root):
  stack = [[ root, 1 ]]
  result = 0

  while stack:
    node, depth = stack.pop()
    
    if node:
      result = max(result, depth)
      stack.append([ node.left, depth + 1 ])
      stack.append([ node.right, depth + 1])

  return result